File: android/app/src/main/java/com/audiobookshelf/app/plugins/TtsPlugin.kt
package com.audiobookshelf.app.plugins

import android.content.Context
import android.os.Bundle
import android.speech.tts.TextToSpeech
import android.speech.tts.UtteranceProgressListener
import android.util.Log
import com.getcapacitor.JSArray
import com.getcapacitor.JSObject
import com.getcapacitor.Plugin
import com.getcapacitor.PluginCall
import com.getcapacitor.PluginMethod
import com.getcapacitor.annotation.CapacitorPlugin
import java.util.Locale

@CapacitorPlugin(name = "TtsPlugin")
class TtsPlugin : Plugin(), TextToSpeech.OnInitListener {

    private val TAG = "TtsPlugin"
    private var tts: TextToSpeech? = null
    private var isReady = false
    private var rate: Float = 1.0f
    private var pitch: Float = 1.0f
    private var selectedVoice: android.speech.tts.Voice? = null

    override fun load() {
        tts = TextToSpeech(context, this)
    }

    override fun onInit(status: Int) {
        if (status == TextToSpeech.SUCCESS) {
            isReady = true
            tts?.language = Locale.getDefault()
            tts?.setOnUtteranceProgressListener(object : UtteranceProgressListener() {
                override fun onStart(utteranceId: String?) {
                    val data = JSObject()
                    data.put("utteranceId", utteranceId)
                    notifyListeners("ttsStart", data)
                }

                override fun onDone(utteranceId: String?) {
                    val data = JSObject()
                    data.put("utteranceId", utteranceId)
                    notifyListeners("ttsDone", data)
                }

                @Deprecated("Deprecated in Java")
                override fun onError(utteranceId: String?) {
                    val data = JSObject()
                    data.put("utteranceId", utteranceId)
                    data.put("error", "unknown")
                    notifyListeners("ttsError", data)
                }

                override fun onRangeStart(utteranceId: String?, start: Int, end: Int, frame: Int) {
                    val data = JSObject()
                    data.put("utteranceId", utteranceId)
                    data.put("start", start)
                    data.put("end", end)
                    notifyListeners("ttsRangeStart", data)
                }
            })
            Log.d(TAG, "TTS initialized successfully")
        } else {
            Log.e(TAG, "TTS initialization failed with status $status")
        }
    }

    @PluginMethod
    fun initialize(call: PluginCall) {
        val result = JSObject()
        result.put("ready", isReady)
        call.resolve(result)
    }

    @PluginMethod
    fun getVoices(call: PluginCall) {
        val voicesArray = JSArray()
        tts?.voices?.forEach { voice ->
            val v = JSObject()
            v.put("name", voice.name)
            v.put("lang", voice.locale.toLanguageTag())
            v.put("localService", !voice.isNetworkConnectionRequired)
            voicesArray.put(v)
        }
        val result = JSObject()
        result.put("voices", voicesArray)
        call.resolve(result)
    }

    @PluginMethod
    fun speak(call: PluginCall) {
        if (!isReady) {
            call.reject("TTS not initialized")
            return
        }
        val text = call.getString("text") ?: run { call.reject("text required"); return }
        val utteranceId = call.getString("utteranceId") ?: "tts_utt"
        val callRate = call.getFloat("rate", rate) ?: rate
        val callPitch = call.getFloat("pitch", pitch) ?: pitch
        val voiceName = call.getString("voiceName")

        tts?.setSpeechRate(callRate)
        tts?.setPitch(callPitch)

        if (voiceName != null) {
            val voice = tts?.voices?.find { it.name == voiceName }
            if (voice != null) tts?.voice = voice
        } else if (selectedVoice != null) {
            tts?.voice = selectedVoice
        }

        val params = Bundle()
        params.putString(TextToSpeech.Engine.KEY_PARAM_UTTERANCE_ID, utteranceId)

        val queueResult = tts?.speak(text, TextToSpeech.QUEUE_FLUSH, params, utteranceId)
        val result = JSObject()
        result.put("started", queueResult == TextToSpeech.SUCCESS)
        call.resolve(result)
    }

    @PluginMethod
    fun stop(call: PluginCall) {
        tts?.stop()
        val result = JSObject()
        result.put("stopped", true)
        call.resolve(result)
    }

    @PluginMethod
    fun pause(call: PluginCall) {
        // Android TTS has no native pause — stop and remember position
        tts?.stop()
        val result = JSObject()
        result.put("paused", true)
        call.resolve(result)
    }

    @PluginMethod
    fun resume(call: PluginCall) {
        // Resume handled at JS layer by re-calling speak from saved index
        val result = JSObject()
        result.put("resumed", true)
        call.resolve(result)
    }

    @PluginMethod
    fun setRate(call: PluginCall) {
        rate = call.getFloat("rate", 1.0f) ?: 1.0f
        tts?.setSpeechRate(rate)
        call.resolve(JSObject().put("ok", true))
    }

    @PluginMethod
    fun setPitch(call: PluginCall) {
        pitch = call.getFloat("pitch", 1.0f) ?: 1.0f
        tts?.setPitch(pitch)
        call.resolve(JSObject().put("ok", true))
    }

    @PluginMethod
    fun setVoice(call: PluginCall) {
        val voiceName = call.getString("voiceName")
        if (voiceName != null) {
            selectedVoice = tts?.voices?.find { it.name == voiceName }
        }
        call.resolve(JSObject().put("ok", true))
    }

    override fun handleOnDestroy() {
        tts?.stop()
        tts?.shutdown()
        tts = null
        super.handleOnDestroy()
    }
}

File: ios/App/App/TtsPlugin.swift
import Foundation
import AVFoundation
import Capacitor
import MediaPlayer

@objc(TtsPlugin)
public class TtsPlugin: CAPPlugin, AVSpeechSynthesizerDelegate {

    private let synthesizer = AVSpeechSynthesizer()
    private var rate: Float = AVSpeechUtteranceDefaultSpeechRate
    private var pitch: Float = 1.0
    private var selectedVoiceIdentifier: String? = nil
    private var currentUtteranceId: String? = nil

    public override func load() {
        synthesizer.delegate = self
        setupAudioSession()
        setupRemoteCommandCenter()
    }

    private func setupAudioSession() {
        do {
            try AVAudioSession.sharedInstance().setCategory(.playback, mode: .spokenAudio, options: [.mixWithOthers])
            try AVAudioSession.sharedInstance().setActive(true)
        } catch {
            print("[TtsPlugin] AVAudioSession setup failed: \(error)")
        }
    }

    private func setupRemoteCommandCenter() {
        let commandCenter = MPRemoteCommandCenter.shared()
        commandCenter.playCommand.addTarget { [weak self] _ in
            self?.notifyListeners("ttsResume", data: [:])
            return .success
        }
        commandCenter.pauseCommand.addTarget { [weak self] _ in
            self?.synthesizer.pauseSpeaking(at: .word)
            self?.notifyListeners("ttsPause", data: [:])
            return .success
        }
        commandCenter.nextTrackCommand.addTarget { [weak self] _ in
            self?.synthesizer.stopSpeaking(at: .word)
            self?.notifyListeners("ttsNext", data: [:])
            return .success
        }
        commandCenter.previousTrackCommand.addTarget { [weak self] _ in
            self?.synthesizer.stopSpeaking(at: .word)
            self?.notifyListeners("ttsPrev", data: [:])
            return .success
        }
    }

    private func updateNowPlaying(title: String = "Read Aloud") {
        var info = [String: Any]()
        info[MPMediaItemPropertyTitle] = title
        info[MPNowPlayingInfoPropertyPlaybackRate] = Double(rate)
        MPNowPlayingInfoCenter.default().nowPlayingInfo = info
    }

    @objc func initialize(_ call: CAPPluginCall) {
        call.resolve(["ready": true])
    }

    @objc func getVoices(_ call: CAPPluginCall) {
        let voices = AVSpeechSynthesisVoice.speechVoices().map { v -> [String: Any] in
            return ["name": v.name, "lang": v.language, "localService": true]
        }
        call.resolve(["voices": voices])
    }

    @objc func speak(_ call: CAPPluginCall) {
        guard let text = call.getString("text") else {
            call.reject("text required")
            return
        }
        let utteranceId = call.getString("utteranceId") ?? "tts_utt"
        let callRate = call.getFloat("rate") ?? rate
        let callPitch = call.getFloat("pitch") ?? pitch
        let voiceName = call.getString("voiceName")

        synthesizer.stopSpeaking(at: .immediate)

        let utterance = AVSpeechUtterance(string: text)
        utterance.rate = callRate
        utterance.pitchMultiplier = callPitch

        if let voiceName = voiceName {
            utterance.voice = AVSpeechSynthesisVoice.speechVoices().first { $0.name == voiceName }
        } else if let identifier = selectedVoiceIdentifier {
            utterance.voice = AVSpeechSynthesisVoice(identifier: identifier)
        }

        currentUtteranceId = utteranceId
        updateNowPlaying()
        synthesizer.speak(utterance)
        call.resolve(["started": true])
    }

    @objc func stop(_ call: CAPPluginCall) {
        synthesizer.stopSpeaking(at: .immediate)
        call.resolve(["stopped": true])
    }

    @objc func pause(_ call: CAPPluginCall) {
        synthesizer.pauseSpeaking(at: .word)
        call.resolve(["paused": true])
    }

    @objc func resume(_ call: CAPPluginCall) {
        if synthesizer.isPaused {
            synthesizer.continueSpeaking()
        }
        call.resolve(["resumed": true])
    }

    @objc func setRate(_ call: CAPPluginCall) {
        rate = call.getFloat("rate") ?? AVSpeechUtteranceDefaultSpeechRate
        call.resolve(["ok": true])
    }

    @objc func setPitch(_ call: CAPPluginCall) {
        pitch = call.getFloat("pitch") ?? 1.0
        call.resolve(["ok": true])
    }

    @objc func setVoice(_ call: CAPPluginCall) {
        if let voiceName = call.getString("voiceName") {
            selectedVoiceIdentifier = AVSpeechSynthesisVoice.speechVoices().first { $0.name == voiceName }?.identifier
        }
        call.resolve(["ok": true])
    }

    // MARK: - AVSpeechSynthesizerDelegate

    public func speechSynthesizer(_ synthesizer: AVSpeechSynthesizer, didStart utterance: AVSpeechUtterance) {
        notifyListeners("ttsStart", data: ["utteranceId": currentUtteranceId ?? ""])
    }

    public func speechSynthesizer(_ synthesizer: AVSpeechSynthesizer, didFinish utterance: AVSpeechUtterance) {
        notifyListeners("ttsDone", data: ["utteranceId": currentUtteranceId ?? ""])
    }

    public func speechSynthesizer(_ synthesizer: AVSpeechSynthesizer, didCancel utterance: AVSpeechUtterance) {
        notifyListeners("ttsError", data: ["utteranceId": currentUtteranceId ?? "", "error": "cancelled"])
    }

    public func speechSynthesizer(_ synthesizer: AVSpeechSynthesizer,
                                   willSpeakRangeOfSpeechString characterRange: NSRange,
                                   utterance: AVSpeechUtterance) {
        notifyListeners("ttsRangeStart", data: [
            "utteranceId": currentUtteranceId ?? "",
            "start": characterRange.location,
            "end": characterRange.location + characterRange.length
        ])
    }
}
